tools/tinygoize/main.go

```diff
@@ -2,81 +2,315 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// usage: invoke this with a list of directories.
-// For each directory, it will try a tinygobuild with
-// CGO_ENABLED=0, GOARCH=amd64, and GOOS=linux.
-// If the tinygo build fails, it will use ast package
-// to rewrite //go:build lines as follows:
-// the line starts as //go:build expr
-// it is rewritten to //go:build !tinygo && (expr)
-// When the file is written, the expression seems
-// to be simplified.
+// usage: invoke this with a list of directories. For each directory, it will
+// run `CGO_ENABLED=0 GOARCH=amd64 GOOS=linux tinygo build -tags tinygo.enable`
+// then attempt to fix-up the build tags by either adding or removing an
+// go:build expression `(!tinygo || tinygo.enable)`
 
 package main
 
 import (
 	"bytes"
 	"flag"
+	"fmt"
+	"go/ast"
 	"go/parser"
 	"go/printer"
 	"go/token"
 	"log"
 	"os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
+	"sort"
 	"strings"
 )
 
 const goBuild = "//go:build "
+const constraint = "!tinygo || tinygo.enable"
 
-func main() {
-	flag.Parse()
+// Additional tags required for specific commands. Assume command names unique
+// despite being in different directories.
+var addBuildTags = map[string]string{
+	"gzip":     "noasm",
+	"insmod":   "noasm",
+	"rmmod":    "noasm",
+	"bzimage":  "noasm",
+	"kconf":    "noasm",
+	"modprobe": "noasm",
+	"console":  "noasm",
+}
+
+func buildTags(dir string) (tags string) {
+	parts := strings.Split(dir, "/")
+	cmd := parts[len(parts)-1]
+	return addBuildTags[cmd]
+}
+
+// Track set of passing, failing, and excluded commands
+type BuildStatus struct {
+	passing  []string
+	failing  []string
+	excluded []string
+}
+
+// return trimmed output of "tinygo version"
+func tinygoVersion(tinygo *string) string {
```

Comment:
I don't think calling `Fatalf` is a clean way to exit here, since it calls `os.Exit(1)` under the hood. Reconsider returning an error and handle this the way up. I see that you did not introduce this but it was there before, but when we do a rework we might as well make it clean, right?

```suggestion
func tinygoVersion(tinygo *string) (string, error) {
```

Reply:
Done.